main.c

```diff
@@ -810,6 +827,9 @@ static void dev_removed(struct tcmu_device *dev)
 	tcmulib_cleanup_cmdproc_thread(dev);
 
 	cleanup_io_work_queue(dev, false);
+
+	if (aio_wait_for_empty_queue(rdev))
```

Comment:
I think we want to move this before tcmulib_cleanup_cmdproc_thread.

If runner is in somewhere in tcmur_generic_handle_cmd (after a track_aio_request_start call, but before the cmd has been sent to the handler), we might cancel the thread before the runner thread calls track_aio_request_finish or before the handler can setup its AIO which would eventually call track_aio_request_finish. We would then wait here forever.

Reply:
Agreed.